pythonx/UltiSnips/store.py

```diff
@@ -0,0 +1,237 @@
+
+class Store(object):
+    """
+    A dict-like supporting supporting the obj[key, default] synthax
+    """
+    def __init__(self):
+        self._dict = dict()
+
+    def __getitem__(self, key):
+        if isinstance(key, tuple) :
+            key, _default = key
+            key = str(key)
+            if key not in self._dict :
+                try :
+                    default = _default()
+                except :
+                    default = _default
+                self[key] = default
+                return default
+        else :
+            key = str(key)
+        return self._dict[key]
+
+    def __getattr__(self, key):
+        try :
+            return getattr(self._dict, key)
+        except :
+            try :
+                return self[key]
+            except :
+                pass
+            raise
+
+    def __setattr__(self, key, val):
+        if key[0] == '_' :
+            return super().__setattr__(key, val)
+        else :
+            self[key] = val
+
+    def __setitem__(self, key, val):
+        key = str(key)
+        return self._dict.__setitem__(key, val)
+    def __delitem__(self, *args, **kwargs):
+        return self._dict.__delitem__(*args, **kwargs)
+    def __len__(self, *args, **kwargs):
+        return self._dict.__len__(*args, **kwargs)
+    def __iter__(self, *args, **kwargs):
+        return self._dict.__iter__(*args, **kwargs)
+    def __iter__(self, *args, **kwargs):
+        return self._dict.__iter__(*args, **kwargs)
+
+    def load(self):
+        pass
+
+    def save(self):
+        pass
+
+
+class BufferStore(Store):
+    """
+    A data store local to a buffer. Any python interpolation in a snippet running the same buffer
+    access the same instance through `snip.store.buffer`
+    """
+    def __init__(self):
+        self._dict = dict()
+
+    _stores = defaultdict(lambda : BufferStore())
+
+    @classmethod
+    def bufferStore(cls, buffer_number):
+        return cls._stores[buffer_number]
+
+
+    @classmethod
+    def currentBufferStore(cls):
+        """
+        Gets or create the store for the current buffer.
+        """
+        return cls.bufferStore(vim_helper.buf.number)
+
+    @classmethod
+    def clean(cls):
+        """
+        Check each buffer if it is alive, else delete it.
+        """
+        keys = set(cls._stores.keys())
+        alive = { b.number for b in vim_helper.vim.buffers if b.valid }
+        for k in keys - alive:
+            del cls._stores[k]
+
+
+class SessionStore(Store):
+    """
+    A Store common to a vim instance. Any python interpolation of any snippet in any buffer can access
+    the same instance through `snip.store.session`
+    """
+    def __init__(self):
+        self._dict = dict()
+
+    _store = None
+    @classmethod
+    def get(cls):
+        if cls._store is None :
+            cls._store = cls()
+        return cls._store
+
+
+class _PersistentStore(Store):
+    """
+    A Store persistent saved to json format.
```

Comment:
this is important information: you can only story stuff that is serializable to JSON. You should notice this in the docs or use 'pickle' to have fewer (but still some) limitations.

Reply:
I agree, I used json because pickle would leave a visious security breach, because one would just have to put malicious code in the common store file to make it executed by vim (and since vim could be used as root by some users...)

I have to better document it